### PR TITLE
PVS warnings: check operands before cast

### PR DESCRIPTION
**Pull request by justinmk**
see https://github.com/neovim/neovim/pull/9465

alternative? https://stackoverflow.com/a/44830670/152142

### CODE REVIEW

File: src/nvim/assert.h [hunk 1]
```diff
@@ -121,4 +121,33 @@
...
+///   https://clang.llvm.org/docs/LanguageExtensions.html
+///   https://gcc.gnu.org/onlinedocs/gcc/Integer-Overflow-Builtins.html
+///
+/// Alternative for compilers without __builtin_xx_overflow ?
+///   https://stackoverflow.com/a/44830670/152142
+///
+/// @param MAX Maximum value of the narrowest type of operand.
+///            Not used if compiler supports __builtin_add_overflow.
+#if (defined(__clang__) && __has_builtin(__builtin_add_overflow)) \
+  || (__GNUC__ >= 5)
```
Comment (justinmk) [inline src/nvim/assert.h]: gcc does not support `__has_builtin`. Is there another way to feature-detect `__builtin_add_overflow` on gcc?

Comment (jamessan) [reply src/nvim/assert.h]: Use CMake's `check_c_source_compiles()` instead?  That would work for both compilers.

File: src/nvim/assert.h [hunk 2]
```diff
@@ -121,4 +121,33 @@
...
+# define STRICT_ADD(a, b, c, t) \
+  do { if (__builtin_add_overflow(a, b, c)) { abort(); } } while (0)
+#else
+# define STRICT_ADD(a, b, c, t) \
+  do { *c = (t)(a + b); } while (0)
```
Comment (jamessan) [inline src/nvim/assert.h]: Why not `*c = (t)(a) + (b)`, which is more inline with PVS' suggestion.

Comment (justinmk) [reply src/nvim/assert.h]: @jamessan Then we run into `-Wsign-conversion` and `-Wconversion` errors. 

So we satisfy PVS and improve correctness on newer systems, but older systems will just need to live with the old approach.

**Comment by justinmk** [comment]
if CI passes, will merge if there are no further comments.

**Comment by justinmk** [comment]
qb fialures look like known flaky tests.

**Pull request merged**